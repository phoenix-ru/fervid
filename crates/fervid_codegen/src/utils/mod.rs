// to_ident_or_str

use std::fmt::{Write, Error};

use fervid_core::{FervidAtom, StrOrExpr};
use swc_core::{common::Span, ecma::ast::{ComputedPropName, Ident, EsReserved, IdentName, PropName, Str}};

/// Adapted from SWC Ident::verify_symbol
#[inline]
pub fn is_valid_ident(s: &str) -> bool {
    if s.is_reserved() || s.is_reserved_in_strict_mode(true) || s.is_reserved_in_strict_bind() {
        return false;
    }

    let mut chars = s.chars();

    if let Some(first) = chars.next() {
        if Ident::is_valid_start(first) && chars.all(Ident::is_valid_continue) {
            return true;
        }
    }

    false
}

pub fn str_to_propname(s: &str, span: Span) -> PropName {
    if is_valid_ident(s) {
        PropName::Ident(IdentName { span, sym: s.into() })
    } else {
        PropName::Str(Str {
            span,
            value: s.into(),
            raw: None,
        })
    }
}

pub fn atom_to_propname(sym: FervidAtom, span: Span) -> PropName {
    if is_valid_ident(&sym) {
        PropName::Ident(IdentName { span, sym })
    } else {
        PropName::Str(Str {
            span,
            value: sym,
            raw: None,
        })
    }
}

pub fn str_or_expr_to_propname(str_or_expr: StrOrExpr, span: Span) -> PropName {
    match str_or_expr {
        StrOrExpr::Str(sym) => atom_to_propname(sym, span),
        StrOrExpr::Expr(expr) => PropName::Computed(ComputedPropName {
            span,
            expr,
        }),
    }
}

pub fn to_camelcase(s: &str, buf: &mut impl Write) -> Result<(), Error> {
    for (idx, word) in s.split('-').enumerate() {
        if idx == 0 {
            buf.write_str(word)?;
            continue;
        }

        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                buf.write_char(ch_component)?;
            }

            // Push the rest of the word
            buf.write_str(&word[ch.len_utf8()..])?;
        }
    }

    Ok(())
}
