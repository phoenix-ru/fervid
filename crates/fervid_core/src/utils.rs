use swc_core::{common::Span, ecma::ast::{ComputedPropName, EsReserved, Ident, IdentName, PropName, Str}};

use crate::{AttributeOrBinding, FervidAtom, StrOrExpr, VBindDirective};

/// Checks whether the attributes name is the same as `expected_name`
#[inline]
pub fn check_attribute_name(attr: &AttributeOrBinding, expected_name: &str) -> bool {
    matches!(attr,
        AttributeOrBinding::RegularAttribute { name, .. } |
        AttributeOrBinding::VBind(VBindDirective { argument: Some(StrOrExpr::Str(name)), .. })
        if name == expected_name
    )
}

/// Adapted from SWC Ident::verify_symbol
#[inline]
pub fn is_valid_ident(s: &str) -> bool {
    if s.is_reserved() || s.is_reserved_in_strict_mode(true) || s.is_reserved_in_strict_bind() {
        return false;
    }

    is_valid_propname(s)
}

#[inline]
pub fn is_valid_propname(s: &str) -> bool {
    let mut chars = s.chars();

    if let Some(first) = chars.next() {
        if Ident::is_valid_start(first) && chars.all(Ident::is_valid_continue) {
            return true;
        }
    }

    false
}

pub fn str_to_propname(s: &str, span: Span) -> PropName {
    if is_valid_propname(s) {
        PropName::Ident(IdentName { span, sym: s.into() })
    } else {
        PropName::Str(Str {
            span,
            value: s.into(),
            raw: None,
        })
    }
}

pub fn atom_to_propname(sym: FervidAtom, span: Span) -> PropName {
    if is_valid_propname(&sym) {
        PropName::Ident(IdentName { span, sym })
    } else {
        PropName::Str(Str {
            span,
            value: sym,
            raw: None,
        })
    }
}

pub fn str_or_expr_to_propname(str_or_expr: StrOrExpr, span: Span) -> PropName {
    match str_or_expr {
        StrOrExpr::Str(sym) => atom_to_propname(sym, span),
        StrOrExpr::Expr(expr) => PropName::Computed(ComputedPropName {
            span,
            expr,
        }),
    }
}
