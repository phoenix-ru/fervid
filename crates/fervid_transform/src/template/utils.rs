use fervid_core::FervidAtom;
use swc_core::{common::DUMMY_SP, ecma::ast::{ArrowExpr, BindingIdent, BlockStmtOrExpr, Expr, Ident, Pat}};

/// `foo-bar-baz` -> `FooBarBaz`
#[inline]
pub(crate) fn to_pascal_case(raw: &str, out: &mut String) {
    for word in raw.split('-') {
        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                out.push(ch_component);
            }

            // Push the rest of the word
            out.push_str(&word[ch.len_utf8()..]);
        }
    }
}

/// `foo-bar-baz` -> `fooBarBaz`
#[inline]
pub(crate) fn to_camel_case(raw: &str, out: &mut String) {
    for (idx, word) in raw.split('-').enumerate() {
        if idx == 0 {
            out.push_str(word);
            continue;
        }

        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                out.push(ch_component);
            }

            // Push the rest of the word
            out.push_str(&word[ch.len_utf8()..]);
        }
    }
}

/// Wraps `expr` to `$event => (expr)`
#[inline]
pub fn wrap_in_event_arrow(expr: Box<Expr>) -> Box<Expr> {
    let evt_param = Pat::Ident(BindingIdent {
        id: Ident {
            span: DUMMY_SP,
            ctxt: Default::default(),
            sym: FervidAtom::from("$event"),
            optional: false,
        },
        type_ann: None,
    });

    Box::new(Expr::Arrow(ArrowExpr {
        span: DUMMY_SP,
        ctxt: Default::default(),
        params: vec![evt_param],
        body: Box::new(BlockStmtOrExpr::Expr(expr)),
        is_async: false,
        is_generator: false,
        type_params: None,
        return_type: None,
    }))
}
