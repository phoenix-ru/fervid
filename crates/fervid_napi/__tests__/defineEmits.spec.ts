import { BindingTypes } from '..'
import { assertCode, compile } from './utils'
import { describe, expect, test } from 'vitest'

describe('defineEmits', () => {
  test('basic usage', () => {
    const { content, bindings } = compile(`
<script setup>
const myEmit = defineEmits(['foo', 'bar'])
</script>
  `, { outputSetupBindings: true })

    assertCode(content)
    expect(bindings).toStrictEqual({
      myEmit: BindingTypes.SETUP_CONST,
    })

    // should remove defineEmits import and call
    expect(content).not.toMatch('defineEmits')

    // should generate correct setup signature
    expect(content).toMatch(
      `setup (__props, { emit: __emit, expose: __expose }) {`,
    )
    expect(content).toMatch('const myEmit = __emit')

    // should include context options in default export
    expect(content).toMatch(`export default {
    emits: [
        'foo',
        'bar'
    ],`)
  })

  test('w/ runtime options', () => {
    const { content } = compile(`
<script setup lang="ts">
const emit = defineEmits(['a', 'b'])
</script>
    `)
    assertCode(content)
//     expect(content).toMatch(`export default /*@__PURE__*/_defineComponent({
//   emits: ['a', 'b'],
//   setup(__props, { expose: __expose, emit: __emit }) {`)

    // TODO PURE
    expect(content).toMatch(`export default _defineComponent({
    emits: [
        'a',
        'b'
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {`)
    expect(content).toMatch('const emit = __emit')
  })

  test('w/ type', () => {
    const { content } = compile(`
    <script setup lang="ts">
    const emit = defineEmits<(e: 'foo' | 'bar') => void>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar"
    ]`)
  })

  test('w/ type (union)', () => {
    const type = `((e: 'foo' | 'bar') => void) | ((e: 'baz', id: number) => void)`
    const { content } = compile(`
    <script setup lang="ts">
    const emit = defineEmits<${type}>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar",
        "baz"
    ]`)
  })

  test('w/ type (type literal w/ call signatures)', () => {
    const type = `{(e: 'foo' | 'bar'): void; (e: 'baz', id: number): void;}`
    const { content } = compile(`
    <script setup lang="ts">
    const emit = defineEmits<${type}>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar",
        "baz"
    ]`)
  })

  test('w/ type (interface)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    interface Emits { (e: 'foo' | 'bar'): void }
    const emit = defineEmits<Emits>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar"
    ]`)
  })

  test('w/ type (interface w/ extends)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    interface Base { (e: 'foo'): void }
    interface Emits extends Base { (e: 'bar'): void }
    const emit = defineEmits<Emits>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "bar",
        "foo"
    ]`)
  })

  test('w/ type (exported interface)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    export interface Emits { (e: 'foo' | 'bar'): void }
    const emit = defineEmits<Emits>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar"
    ]`)
  })

  test('w/ type from normal script', () => {
    const { content } = compile(`
    <script lang="ts">
      export interface Emits { (e: 'foo' | 'bar'): void }
    </script>
    <script setup lang="ts">
    const emit = defineEmits<Emits>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar"
    ]`)
  })

  test('w/ type (type alias)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    type Emits = { (e: 'foo' | 'bar'): void }
    const emit = defineEmits<Emits>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar"
    ]`)
  })

  test('w/ type (exported type alias)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    export type Emits = { (e: 'foo' | 'bar'): void }
    const emit = defineEmits<Emits>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar"
    ]`)
  })

  test('w/ type (referenced function type)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    type Emits = (e: 'foo' | 'bar') => void
    const emit = defineEmits<Emits>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar"
    ]`)
  })

  test('w/ type (referenced exported function type)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    export type Emits = (e: 'foo' | 'bar') => void
    const emit = defineEmits<Emits>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar"
    ]`)
  })

  // #5393
  test('w/ type (interface ts type)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    interface Emits { (e: 'foo'): void }
    const emit: Emits = defineEmits(['foo'])
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        'foo'
    ]`)
  })

  test('w/ type (property syntax)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    const emit = defineEmits<{ foo: [], bar: [] }>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo",
        "bar"
    ]`)
  })

  // #8040
  test('w/ type (property syntax string literal)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    const emit = defineEmits<{ 'foo:bar': [] }>()
    </script>
    `)
    assertCode(content)
    expect(content).toMatch(`emits: [
        "foo:bar"
    ]`)
  })

  // #7943
  test('w/ type (type references in union)', () => {
    const { content } = compile(`
    <script setup lang="ts">
    type BaseEmit = "change"
    type Emit = "some" | "emit" | BaseEmit
    const emit = defineEmits<{
      (e: Emit): void;
      (e: "another", val: string): void;
    }>();
    </script>
    `)

    assertCode(content)
    expect(content).toMatch(`emits: [
        "another",
        "some",
        "emit",
        "change"
    ]`)
  })

  describe('errors', () => {
    test('w/ both type and non-type args', () => {
      const { errors } = compile(`<script setup lang="ts">
        defineEmits<{}>({})
        </script>`)

      expect(errors.length).toBe(1)
      expect(errors[0].message).toMatch('DefineEmitsTypeAndNonTypeArguments')
    })

    test('mixed usage of property / call signature', () => {
      const { errors } = compile(`<script setup lang="ts">
        defineEmits<{
          foo: []
          (e: 'hi'): void
        }>()
        </script>`)

      expect(errors.length).toBe(1)
      expect(errors[0].message).toMatch('DefineEmitsMixedCallAndPropertySyntax')
    })
  })
})
