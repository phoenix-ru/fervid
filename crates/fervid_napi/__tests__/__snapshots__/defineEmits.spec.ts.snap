// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineEmits > basic usage 1`] = `
"export default {
    emits: [
        'foo',
        'bar'
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const myEmit = __emit;
        return {
            myEmit
        };
    }
};
"
`;

exports[`defineEmits > w/ runtime options 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        'a',
        'b'
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (exported interface) 1`] = `
"export interface Emits {
    (e: 'foo' | 'bar') : void;
}
import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (exported type alias) 1`] = `
"export type Emits = {
    (e: 'foo' | 'bar') : void;
};
import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (interface ts type) 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        'foo'
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        interface Emits {
            (e: 'foo') : void;
        }
        const emit: Emits = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (interface w/ extends) 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "bar",
        "foo"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        interface Base {
            (e: 'foo') : void;
        }
        interface Emits extends Base {
            (e: 'bar') : void;
        }
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (interface) 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        interface Emits {
            (e: 'foo' | 'bar') : void;
        }
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (property syntax string literal) 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo:bar"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (property syntax) 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (referenced exported function type) 1`] = `
"export type Emits = (e: 'foo' | 'bar') => void;
import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (referenced function type) 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        type Emits = (e: 'foo' | 'bar') => void;
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (type alias) 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        type Emits = {
            (e: 'foo' | 'bar') : void;
        };
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (type literal w/ call signatures) 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar",
        "baz"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (type references in union) 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "another",
        "some",
        "emit",
        "change"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        type BaseEmit = "change";
        type Emit = "some" | "emit" | BaseEmit;
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type (union) 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar",
        "baz"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type 1`] = `
"import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;

exports[`defineEmits > w/ type from normal script 1`] = `
"export interface Emits {
    (e: 'foo' | 'bar') : void;
}
import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    emits: [
        "foo",
        "bar"
    ],
    __name: "anonymous",
    setup (__props, { emit: __emit, expose: __expose }) {
        __expose();
        const emit = __emit;
        return {
            emit
        };
    }
});
"
`;
